### PR TITLE
Register debts upon withdrawals

### PR DESCRIPTION
**Pull request by larryob**
BaseVault/AffineVault TODO

- [x] Consider not allowing pendingDebt to be deposited into strategies
- [x] Move pendingDebt into BaseVault and AffineVault
- [x] Add custom divestment data to allow vault to forcefully liquidate a strategy (removeStrategy)
- [x] Add soft liquidations (used in cross-chain/internal rebalancing) => vault does not forcefully liquidate and is issued its own debtShares
  - [x] Vault must loop through strategies after attempted liquidation is complete and assign debt proportionally to each strategy
- [x] Implement redeemOwnShares(vault redeems debtShares from escrow) and settleStrategyDebt
- [x] Add tests

Vault/L2Vault TODO
- [x] Integrate with actual escrow contract
- [x] Account for pendingDebt during withdrawals => pendingDebt can’t be withdrawn
- [x] Issue debt shares to vault if needed in `_l1l2Rebalance`
- [x] Issue debt shares to user  if needed during withdrawals

L1Vault TODO
- [x] Issue debt shares to vault during cross-chain rebalancing (`processFundRequest`)

### CODE REVIEW

**Comment by larryob** [comment]
Fixes ALP-4041

File: src/vaults/AffineVault.sol [hunk 1]
```diff
@@ -424,33 +432,75 @@ contract AffineVault is AffineGovernable, AccessControlUpgradeable {
...
+        for (uint256 i = 0; i < MAX_STRATEGIES; i = uncheckedInc(i)) {
+            Strategy strategy = withdrawalQueue[i];
+            if (address(strategy) == address(0)) break;
+            if (responses[i] != DivestResponse.DEBT) continue;
+
+            uint16 stratBps = strategies[strategy].tvlBps;
+            uint256 debt = (debtToCreate * stratBps) / debtStrategyBps;
+            debtToCreate -= debt;
+            debtStrategyBps -= stratBps;
+            strategy.increaseDebt(debt);
```
Comment (maruf0011) [inline src/vaults/AffineVault.sol:483]: Assigning debt to all strategy after doing possible withdrawal might end up empty strategy  having a debt. We should only assign debt to strategy which fails to withdraw or withdraw less than expected.

File: src/vaults/Vault.sol [hunk 2]
```diff
@@ -125,8 +128,23 @@ contract Vault is AffineVault, ERC4626Upgradeable, PausableUpgradeable, Detailed
...
-        assets = Math.min(_asset.balanceOf(address(this)), assets);
+        // If user tries to withdraw too much, register request on LockedWithdrawalEscrow
+        uint256 currAssets = _asset.balanceOf(address(this));
+        if (currAssets < assets) {
+            uint256 sharesToLock = _convertToShares(assets - currAssets, MathUpgradeable.Rounding.Up);
+            // Lock users shares in escrow
+            _transfer({from: owner, to: address(debtEscrow), amount: sharesToLock});
+
+            // Register withdrawal request
+            debtEscrow.registerWithdrawalRequest(_msgSender(), sharesToLock);
```
Comment (maruf0011) [inline src/vaults/Vault.sol:139]: we need to register debt of `assets - currAssets`. cause debt token = shares* price.

File: src/vaults/Vault.sol [hunk 3]
```diff
@@ -236,13 +254,24 @@ contract Vault is AffineVault, ERC4626Upgradeable, PausableUpgradeable, Detailed
...
+    function clearLockedShares() external onlyRole(HARVESTER) {
+        // Given a certain amount of pendingDebt, settle the debt by burning the equivalent amount of locked shares
+        // and sending the pendingDebt to debtEscrow
+        uint256 sharesToBurn = _convertToShares(pendingDebt, MathUpgradeable.Rounding.Up);
+        _burn(address(debtEscrow), sharesToBurn);
+
+        // Let debtEscrow know that we have settled a debt
+        _asset.transfer(address(debtEscrow), pendingDebt);
+        pendingDebt = 0;
+        debtEscrow.resolveDebtShares(sharesToBurn);
```
Comment (maruf0011) [inline src/vaults/Vault.sol:274]: Resolve share will get the `pending_debt * ration of burned shares`

File: src/vaults/cross-chain-vault/BaseVault.sol [hunk 4]
```diff
@@ -478,33 +487,75 @@ abstract contract BaseVault is AccessControlUpgradeable, AffineGovernable, Multi
...
-            amountLiquidated += withdrawn;
+        // If we liquidated enough OR there are no strategies with debt, exit
+        if (assetsRequested <= currAssets || debtStrategyBps == 0) {
+            return (assetsLiquidated, debtCreated);
         }
-        emit Liquidation({assetsRequested: amount, assetsLiquidated: amountLiquidated});
-        return amountLiquidated;
+
+        // Issue debt to strategies which failed to liquidate. Debt is issued proportionally according to assigned tvl bps
+        uint256 debtToCreate = assetsRequested - currAssets;
```
Comment (maruf0011) [inline src/vaults/cross-chain-vault/BaseVault.sol:527]: Repetitive code, Can we merge it into single vault.

**Pull request closed without merging**